### PR TITLE
feat(wrappers): add decoration wrapper

### PR DESCRIPTION
**Pull request by Dam-Buty**
To test this PR : 
* check it out locally
* run the Literal Api on this branch : damien/eng-1724-integrations-change-how-we-handle-metadatatags
* unskip all sections marked with "Skip for the CI"
* run the tests with a .env containing an OPENAI_API_KEY.

# Changes

* API : add possibility to attach a generation to an existing step (unused in the SDK but exposed to users)

```ts
await literalClient.api.createGeneration(generation, stepId)
```

* OpenAI : 
  * fix option to add metadata/tags at the call level
  * add option to specify a Step ID at the call level
  * fix an issue when the same OpenAI client is instrumented more than once (mostly useful for the tests)

```ts
const openai_ = new OpenAI();

const openai = client.instrumentation.openai({
  // the initial client needs to be passed when instrumenting
  client: openai_
});

await openai.chat.completions.create(
  {
    model: 'gpt-3.5-turbo',
    messages: [
      { role: 'system', content: 'You are a helpful assistant.' },
      { role: 'user', content: 'What is the capital of Canada?' }
    ]
  },
  {
    literalaiTags: ['tag3', 'tag4'],
    literalaiMetadata: { otherKey: 'otherValue' },
    // You can also specify a Step ID at the call level
    // When this generation is logged, it will be created with the given Step ID
    literalaiStepId
  }
);
```

* Langchain
  * add an option to specify a Step ID in a call's metadata
  * add tags/metadata from LangChain to Literal AI generations

```ts
const literalaiStepId = uuidv4();

await model.invoke('Hello, how are you?', {
  callbacks: [cb],
  metadata: {
    key: 'value',
    // use literalaiStepId in the metadata to specify a Step ID
    literalaiStepId,
  },
  tags: ['tag1', 'tag2'],
});
```

* Vercel AI SDK : add option to add tags, metadata and a step ID at the call level

```ts
const { text } = await generateText({
  model: openai('gpt-3.5-turbo'),
  prompt: question,
  literalaiStepId,
  literalaiTags: ['tag1', 'tag2'],
  literalaiMetadata: { otherKey: 'otherValue' }
});
```

* Decorate wrapper : this wrapper allows you to store tags, metadata and a step ID inside the context
  * Threads, steps and generations logged inside the context will inherit the tags & metadata
  * The first generation/step created inside the context will inherit the step ID

```ts
await client.decorate({ metadata, tags, stepId }).wrap(async () => {
  // This generation will be logged with the provided stepId, metadata and tags
  const completion = await openai.chat.completions.create({
    model: 'gpt-4',
    messages: [{ role: 'user', content: 'Say hello !' }]
  });

  // Because step IDs are unique, this call will revert to the default behaviour of
  // assigning a random UUID to the generation when it is logged.
  const completion = await openai.chat.completions.create({
    model: 'gpt-4',
    messages: [{ role: 'user', content: 'Say hello !' }]
  });
})
```

### CODE REVIEW

File: openai.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import 'dotenv/config';
```
Comment (willydouhard) [inline openai.ts:1]: this looks like a test file?

File: package.json [hunk 2]
```diff
@@ -52,6 +52,7 @@
     "zod-to-json-schema": "^3.23.0"
   },
   "dependencies": {
+    "@langchain/openai": "^0.2.7",
```
Comment (willydouhard) [inline package.json:55]: this should be a peer dependency

File: prompt.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import 'dotenv/config';
```
Comment (willydouhard) [inline prompt.ts:1]: this looks like a test file

File: src/instrumentation/langchain.ts [hunk 4]
```diff
@@ -375,97 +405,116 @@ export class LiteralCallbackHandler extends BaseCallbackHandler {
...
+          await this.steps[runId].send();
+        } else {
+          await this.client.api.createGeneration({
+            ...generation,
+            id: this.getGenerationStepId(metadata)
+          });
+        }
       }
+    } catch (e) {
+      console.error(e);
```
Comment (willydouhard) [inline src/instrumentation/langchain.ts:516]: one log should suffice

File: tests/api.test.ts [hunk 5]
```diff
@@ -119,6 +123,9 @@ describe('End to end tests for the SDK', function () {
 
     await client.api.deleteThread(thread.id);
 
+    // We have to await 5 seconds for the thread to disappear from the cache
```
Comment (willydouhard) [inline tests/api.test.ts:126]: related to the cache PR but I think we should get rid of it

File: openai.ts (hunk 1 above)
Comment (clementsirieix) [reply openai.ts]: Done
Moved to an example folder

File: package.json (hunk 2 above)
Comment (clementsirieix) [reply package.json]: Done

File: prompt.ts (hunk 3 above)
Comment (clementsirieix) [reply prompt.ts]: Done

File: src/instrumentation/langchain.ts (hunk 4 above)
Comment (clementsirieix) [reply src/instrumentation/langchain.ts]: Done

File: tests/api.test.ts (hunk 5 above)
Comment (clementsirieix) [reply tests/api.test.ts:123]: Done

**Review by willydouhard — APPROVED** [review]

**Pull request merged**